### PR TITLE
Add linter to check if shuffle before sharding

### PR DESCRIPTION
**Pull request by ejguan**
This PR introduces the linter function to validate there is a shuffle operation before sharding. (Required by TorchText in https://github.com/pytorch/text/pull/1729)

- When `sharding_filter` is not presented in the graph, this function always returns `True`
- For single-path graph, `shuffle` needs to be placed before `sharding_filter`.
- For multi-path graph, any `sharding_filter` requires a `shuffle` before along the path

This linter function won't check if there are multiple `sharding_filter` in the graph or `sharding_filter` is at the right place

### CODE REVIEW

File: torchdata/dataloader2/linter.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+from torchdata.datapipes.iter import IterDataPipe, ShardingFilter, Shuffler
+
+from ._graph_utils import DataPipeGraph, traverse
+
+
+def _check_shuffle_before_sharding(datapipe: IterDataPipe) -> bool:
```
Comment (ejguan) [inline torchdata/dataloader2/linter.py:13]: Any suggestion on the function name is welcomed. :P

**Review by NivekT — COMMENTED** [review]
Are we planning to expose some API to the public? Some options that come to my mind:

1. Users can directly import and call `linter` as a function
    ```python
    dp = dp.map(...).filter(...)
    linter(dp)
    ```
2. No public API, but DataLoader V2 will use call `linter` internally
    - It may be useful for users to be able to enable/disable certain rules
3. Both

File: test/test_linter.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+def dummy_fn(x):
+    return x
+
+
+class LinterTest(unittest.TestCase):
+    def test_sharding_shuffle(self):
+        source_dp = IterableWrapper(list(range(20)))
+
+        # Single path
```
Comment (parmeet) [inline test/test_linter.py:23]: Should we also test for False assertion when sharding is applied before?

```python
        dp = source_dp.map(dummy_fn).sharding_filter().shuffle()
        self.assertFalse(_check_shuffle_before_sharding(dp))
```

**Review by parmeet — APPROVED** [review]
Overall LGTM! Thanks for providing this functionality. I think vision might be interested to use this as well cc: @NicolasHug

**Comment by ejguan** [comment]
> Are we planning to expose some API to the public? Some options that come to my mind:
> 
> 1. Users can directly import and call `linter` as a function
>    ```python
>    dp = dp.map(...).filter(...)
>    linter(dp)
>    ```
> 2. No public API, but DataLoader V2 will use call `linter` internally
>    
>    * It may be useful for users to be able to enable/disable certain rules
> 3. Both

I prefer this function being private for now. I think there are multiple linter functions needed. I would think aggregating all these linter functions together to provide a common API like `linter(dp, sharding_before_shuffle=True, ...)` with options to turn on each specific functionalities for users.

**Review by NivekT — APPROVED** [review]
LGTM!

**Comment by VitalyFedyunin** [comment]
Agree on keeping it private. Eventually, we want to have it on auto in DLv2, but let's navigate the landscape and figure out if we want it on/off by default and if it is applicable for all Reading services or should be reading service level linters.

**Comment by pmeier** [comment]
> This PR introduces the linter function to validate there is a shuffle operation before sharding

Is this the right objective? Shouldn't the objective be "validate that the shuffle operation comes before the operation *iff a shuffle operation is part of the graph*"?

Otherwise something like this will trip the linter:

```py
from torchdata.datapipes.iter import ShardingFilter, IterableWrapper
from torchdata.dataloader2.linter import _check_shuffle_before_sharding

dp = IterableWrapper([])
dp = ShardingFilter(dp)

assert _check_shuffle_before_sharding(dp)
```

**Comment by ejguan** [comment]
> > This PR introduces the linter function to validate there is a shuffle operation before sharding
> 
> Is this the right objective? Shouldn't the objective be "validate that the shuffle operation comes before the sharding operation _iff a shuffle operation is part of the graph_"?
> 
> Otherwise something like this will trip the linter:
> 
> ```python
> from torchdata.datapipes.iter import ShardingFilter, IterableWrapper
> from torchdata.dataloader2.linter import _check_shuffle_before_sharding
> 
> dp = IterableWrapper([])
> dp = ShardingFilter(dp)
> 
> assert _check_shuffle_before_sharding(dp)
> ```

@pmeier 
I thought we always want to have a `shuffle` for domain library. Then, we can provide the right way to run shuffle and sharding. For users who don't want to enable `shuffle`, they still can specify shuffle to False.

**Comment by pmeier** [comment]
> I thought we always want to have a `shuffle` for domain library.

That is true, but this is not a linter for domain libraries, but a general linter applied by the data loader, right? If not, you can disregard my comments.

**Comment by ejguan** [comment]
> > I thought we always want to have a `shuffle` for domain library.
> 
> That is true, but this is not a linter for domain libraries, but a general linter applied by the data loader, right? If not, you can disregard my comments.

@pmeier 
I would like to keep it as it is. I added this PR simply because TorchText team asked me to add this functionality to verify their implementation.

For user-facing linter, I would rely on this #364 to tack the features. Will add your comment to the issue.
